### PR TITLE
Runtime config refactors

### PR DESCRIPTION
**Pull request by rcoh**
## Motivation and Context
Part of refactoring config bag to work with the new RuntimePlugin API

## Description
- change config bag to a vector instead of a linked list

## Testing
- existing UTs

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

**Review by ysaito1001 — APPROVED** [review]

File: rust-runtime/aws-smithy-types/src/config_bag.rs [hunk 1]
```diff
@@ -77,7 +78,7 @@ impl<T: Default> Default for Value<T> {
 }
 
 /// A named layer comprising a config bag
-struct Layer {
+pub struct Layer {
```
[the remark this answers is not in the document]
Comment (rcoh) [reply rust-runtime/aws-smithy-types/src/config_bag.rs:81]: service config needs to use Frozen layer

File: rust-runtime/aws-smithy-types/src/config_bag.rs [hunk 2]
```diff
@@ -192,110 +193,59 @@ impl Debug for Layer {
...
-    pub fn base() -> Self {
-        ConfigBag {
-            head: Layer {
-                name: Cow::Borrowed("base"),
-                props: Default::default(),
-            },
-            tail: None,
-        }
+    /// Insert `value` into the bag
+    pub fn put<T: Send + Sync + Debug + 'static>(&mut self, value: T) -> &mut Self {
```
[the remark this answers is not in the document]
Comment (rcoh) [reply rust-runtime/aws-smithy-types/src/config_bag.rs:238]: good call—I'll fix the docs. this put impl should be deprecated—it's only to support putting non-storable types

File: rust-runtime/aws-smithy-types/src/config_bag.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ysaito1001) [reply rust-runtime/aws-smithy-types/src/config_bag.rs:81]: > service config needs to use Frozen layer

Ah, that's right. A service config will own a `FrozenLayer`, which is cloneable by virtue of `Arc`. It's just that if a service config _builder_ owns a `Layer`, where the builder currently derives `Clone` and `Default`, `Layer` needs to implement those traits as well 🤔

[the remark this answers is not in the document]
Comment (rcoh) [reply rust-runtime/aws-smithy-types/src/config_bag.rs:81]: hmm...you can't clone a layer by virtue of the things inside of it. I wonder if the builder should have the individual fields inside of it and only convert to layer on `.build()`

File: rust-runtime/aws-smithy-types/src/config_bag.rs [hunk 3]
```diff
@@ -429,50 +466,42 @@ impl ConfigBag {
...
-    pub fn freeze(self) -> FrozenConfigBag {
-        self.into()
-    }
-
     /// Add another layer to this configuration bag
     ///
     /// Hint: If you want to re-use this layer, call `freeze` first.
     /// ```
-    /// use aws_smithy_types::config_bag::ConfigBag;
+    /// /*
```
Comment (jdisanti) [inline rust-runtime/aws-smithy-types/src/config_bag.rs:473]: Did you mean to comment out this code example?

**Pull request merged**